components/usage/pkg/db/workspace_instance_usage_test.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package db_test
+
+import (
+	"context"
+	"database/sql"
+	"testing"
+	"time"
+
+	"github.com/gitpod-io/gitpod/usage/pkg/db"
+	"github.com/gitpod-io/gitpod/usage/pkg/db/dbtest"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCanCreateUsageRecords(t *testing.T) {
+	teamID := uuid.New().String()
+	teamAttributionID := db.NewTeamAttributionID(teamID)
+	ctx := context.Background()
+
+	scenarios := []struct {
+		Name         string
+		UsageRecords []db.WorkspaceInstanceUsage
+	}{
+		{
+			Name: "One workspace instance",
+			UsageRecords: []db.WorkspaceInstanceUsage{{
+				WorkspaceID:   uuid.New(),
+				AttributionID: teamAttributionID,
+				StartedAt:     time.Now(),
+				StoppedAt:     sql.NullTime{},
+				CreditsUsed:   0,
+				GenerationId:  0,
+				Deleted:       false,
+			}},
+		},
+		{
+			Name: "Two different workspace instances",
+			UsageRecords: []db.WorkspaceInstanceUsage{{
+				WorkspaceID:   uuid.New(),
+				AttributionID: teamAttributionID,
+				StartedAt:     time.Now(),
+				StoppedAt:     sql.NullTime{},
+				CreditsUsed:   0,
+				GenerationId:  0,
+				Deleted:       false,
+			}},
+		},
+	}
+
+	for _, scenario := range scenarios {
+		conn := dbtest.ConnectForTests(t)
+		t.Run(scenario.Name, func(t *testing.T) {
+			t.Cleanup(func() {
+				require.NoError(t, conn.Where("1=1").Delete(&db.WorkspaceInstanceUsage{}).Error)
```

Comment:
This was the cause of the problems with flaky tests. When more tests exist and do this, it deletes the data needed for other runs of the test. That's why the current approach adds a hook for each ID created to only deleted the IDs created in each test.

Reply:
🤔 

None of these tests call `t.Parallel`, so AIUI the tests in this file will run in series, and within each test each subtest will also run in series. So where is the risk of having a subtest clearing the table after it finishes?

Is it tests running in different packages (which will be run in parallel by `go test`) that is the concern here?

Reply:
https://github.com/gitpod-io/gitpod/pull/10971 has extra context on this, including a link to some literature

Reply:
> But tests in different files will be run concurrently.

I don't think that's correct. See for example [here](https://engineering.mercari.com/en/blog/entry/20220408-how_to_use_t_parallel/):

> By default, execution of test code using the testing package will be done sequentially. However, note that it is only the tests within a given package that run sequentially.

> If tests from multiple packages are specified, the tests will be run in parallel at the package level. For example, imagine there are two packages, package a and package b. The test code in package a will be run sequentially, and the test code in package b will also be run sequentially. However, the tests for package a and package b will be run in parallel. Let’s look more closely into how these will be run in parallel.

So I guess I'm struggling to understand how test cleanup code within a package can interfere with other tests in the same package if there is no `t.Parallel` anywhere. Maybe there is a race in actually deleting the rows from the database?

Reply:
The issues we observed were across different packages. And the same will happen here once we add proper tests to the controller package.

Reply:
test cleanup changed to only remove records that were created by each test.